### PR TITLE
Improve Font Symbols

### PR DESCRIPTION
**Pull request by b4l**
This handles fonts as marks and some other small fixes

@mholthausen could you have a look at this? Somehow I can not request a review from you. Thanks!

### CODE REVIEW

File: src/MapfileStyleParser.ts [hunk 1]
```diff
@@ -362,12 +361,18 @@ export class MapfileStyleParser implements StyleParser {
...
+      markSymbolizer.fillOpacity = mapfileStyle.symbol.filled.toLowerCase() === 'true' ? 1 : 0;
+    } else {
+      markSymbolizer.fillOpacity = 0;
+    }
 
     if (mapfileStyle.size) {
       markSymbolizer.radius = mapfileStyle.size / 2;
     }
 
+    markSymbolizer.rotate = mapfileStyle.angle ? mapfileStyle.angle * 1 : 0;
```
Comment (jansule) [inline src/MapfileStyleParser.ts:374]: What is the benefit of multiplying the angle by 1?

File: src/MapfileStyleParser.ts [hunk 2]
```diff
@@ -382,33 +387,48 @@ export class MapfileStyleParser implements StyleParser {
...
-      }
-      markSymbolizer.wellKnownName = 'circle' as WellKnownName;
-    } else {
-      if (isSquare(points)) {
-        markSymbolizer.wellKnownName = 'square' as WellKnownName;
+    if (mapfileStyle.symbol.points) {
+      const points = mapfileStyle.symbol.points.split(' ').map((item) => parseFloat(item));
+      if (symbolType === 'ellipse') {
+        if (!(points[0] === points[1] && points.length === 2)) {
+          console.warn('Custom ellipse not supported by MarkerSymbolyzer, fallback to "Circle"!');
```
Comment (jansule) [inline src/MapfileStyleParser.ts:394]: I'm not sure, if the parser should provide such fallbacks, or if this is rather a project specific post processing step

File: src/MapfileStyleParser.ts [hunk 3]
```diff
@@ -382,33 +387,48 @@ export class MapfileStyleParser implements StyleParser {
...
+        }
+        if (isCross(points)) {
+          markSymbolizer.wellKnownName = 'cross' as WellKnownName;
+        }
       }
-      if (isTriangle(points)) {
-        markSymbolizer.wellKnownName = 'triangle' as WellKnownName;
+      if (!markSymbolizer.wellKnownName) {
+        console.warn(
+          `Custom symbol not supported by MarkerSymbolyzer, fallback to 'X':\n${JSON.stringify(
```
Comment (jansule) [inline src/MapfileStyleParser.ts:410]: see comment above

File: src/MapfileStyleParser.ts [hunk 4]
```diff
@@ -382,33 +387,48 @@ export class MapfileStyleParser implements StyleParser {
...
+            mapfileStyle.symbol,
+            null,
+            2
+          )}`
+        );
+        markSymbolizer.wellKnownName = 'X' as WellKnownName;
```
Comment (jansule) [inline src/MapfileStyleParser.ts:416]: see comment above

File: src/MapfileStyleParser.ts [hunk 5]
```diff
@@ -526,9 +563,10 @@ export class MapfileStyleParser implements StyleParser {
 
     if (typeof mapfileStyle.symbol === 'object') {
       const symbolType = mapfileStyle.symbol.type;
-      switch (symbolType) {
+      switch (symbolType.toLowerCase()) {
```
Comment (jansule) [inline src/MapfileStyleParser.ts:566]: the indentation seems off here

**Review by jansule — COMMENTED** [review]
This looks quite good! There a just a few minor things to address.

Btw: Congratulations for the great presentation at FOSSGIS today!

File: src/MapfileStyleParser.ts (hunk 2 above)
Comment (b4l) [reply src/MapfileStyleParser.ts:394]: It depends on how much gets preserved of the style without fallback. Probably it is not necessary, though nice to have a visible indication and a dummy style to start off post processing.

File: src/MapfileStyleParser.ts (hunk 5 above)
Comment (b4l) [reply src/MapfileStyleParser.ts:566]: if i format the switch cases get intended, which result in a lint error.

File: src/MapfileStyleParser.ts (hunk 4 above)
Comment (KaiVolland) [reply src/MapfileStyleParser.ts:416]: All WellKnownNames are now lowercase so this should be `'x'`.

File: src/MapfileStyleParser.ts (hunk 2 above)
Comment (KaiVolland) [reply src/MapfileStyleParser.ts:394]: 1. I think it depends on the users preferences what should happen if some style-features are not supported. IMHO we should not transform here.
2. This case is kind of special though. As it effectively is a circle. We're back at the discussion [file-similarity vs visual-similarity](https://github.com/geostyler/geostyler/discussions/1451). Cause we basically want to be able to parse forth and back and still receive the same result as the original input.
3. We could introduce `ellipse` as WellKnownName to the geostyler-style. Which would resolve both issues here, right? If there are even more "basic" styles we dont support yet, we can add them, too. The geostyler-style is not bound to any official standard here as it is just an exchange format.

Comment (b4l) [reply src/MapfileStyleParser.ts:394]: @KaiVolland this sounds reasonable, i don't really mind any direction here. There is just that an ellipse in geostyler style is not sufficient to handle all custom vector symbols form mapserver.

Comment (KaiVolland) [reply src/MapfileStyleParser.ts:394]: Im not into mapserver styling at all. But the vector symbols seem to be kind of a special thing of mapserver. They wouldn't be to hard to parse and i can imagine that you could create similar symbols for other styles too (canvas, base64svg, …), but im not sure if this is worth the effort at this point.

Comment (b4l) [reply src/MapfileStyleParser.ts:394]: Yes, we were thinking about handling them as svg. Never got the time though. This pr is basically my hacks i did shortly before the presentation to the customer.

File: src/MapfileStyleParser.ts (hunk 1 above)
Comment (mholthausen) [reply src/MapfileStyleParser.ts]: `ANGLE`s that are specified in MapFiles with `AUTO`, `AUTO2` or `FOLLOW` instead of an explicit value are returned with `NaN` at this point. Is this considered? The `ANGLE` of `LABEL` allow `[double|auto|auto2|follow|attribute]`

Multiplying by 1, if an explicit value is present as a string, would generate a floating point number from it in any case.

File: src/MapfileStyleParser.ts (hunk 2 above)
Comment (mholthausen) [reply src/MapfileStyleParser.ts:394]: I think introducing such a fallback or dummy style to start off post processing would be quite helpful, for instance as `WellKnownName`. This would certainly simplify it for all parsers to deal with such cases. Project specific adjustments can then still be done in post processing. In my opinion, the prerequisite would be that it can then be implemented in all other parsers as well.

Comment (b4l) [reply src/MapfileStyleParser.ts:394]: Has there been a consent on this? I would like to finish this one way or the other if possible. Thanks!

Comment (jansule) [reply src/MapfileStyleParser.ts:394]: I would suggest to not provide a fallback here. So you could follow @KaiVolland's recommendation of either adding the `ellipse` type to our list of WellKnownNames and then create the proper style object, or handle such cases in a post-processing step.

**Review by jansule — APPROVED** [review]
Very nice work @b4l!
